File: java/log-levels/src/main/java/LogLevels.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LogLevels {

    public static String message(String logLine) throws IllegalArgumentException {
        Pattern pattern = Pattern.compile("^\\[(?:INFO|WARNING|ERROR)\\]: (?<message>.+)");
        Matcher matcher = pattern.matcher(logLine);
        if (matcher.find()) {
            return matcher.group("message").trim();
        } else {
            throw new IllegalArgumentException("Bad logLine");
        }
    }

    public static String logLevel(String logLine) throws IllegalArgumentException {
        Pattern pattern = Pattern.compile("^\\[(?<level>(?:INFO|WARNING|ERROR))\\]: .+");
        Matcher matcher = pattern.matcher(logLine);
        if (matcher.find()) {
            return matcher.group("level").toLowerCase();
        } else {
            throw new IllegalArgumentException("Bad logLine");
        }
    }

    public static String reformat(String logLine)  throws IllegalArgumentException {
        String message = LogLevels.message(logLine);
        String logLevel = LogLevels.logLevel(logLine);
        return String.format("%s (%s)", message, logLevel);
    }

}

File: clojure/log-levels/src/log_levels.clj
(ns log-levels
  (:require [clojure.string :as str]))

(defn message
  "Given a string representing a log line,
  return its message with whitespace trimmed."
  [log-line]
  (let [[_ msg] (str/split log-line #":" 2)]
    (str/trim msg)))

(defn log-level
  "Given a string representing a log line,
   return its level in lower-case."
  [log-line]
  (let [[_ level] (re-find #"\[(\w+)\]" log-line)]
    (str/lower-case level)))

(defn reformat
  "Given a string representing a log line,
  format it with the message first and the log level in parentheses."
  [log-line]
  (let [msg (message log-line)
        level (log-level log-line)]
    (format "%s (%s)" msg level)))
